ext/pdo_oci/oci_statement.c

```diff
@@ -793,15 +793,197 @@ static int oci_stmt_get_col(pdo_stmt_t *stmt, int colno, char **ptr, size_t *len
+			add_assoc_string(return_value, "native_type", "RAW");
+			break;
+		case SQLT_LBI:
+			add_assoc_string(return_value, "oci:decl_type", "LONG RAW");
+			add_assoc_string(return_value, "native_type", "LONG RAW");
+			break;
+		case SQLT_CHR:
+			add_assoc_string(return_value, "oci:decl_type", "VARCHAR2");
+			add_assoc_string(return_value, "native_type", "VARCHAR2");
+			break;
+		case SQLT_AFC:
+			add_assoc_string(return_value, "oci:decl_type", "CHAR");
+			add_assoc_string(return_value, "native_type", "CHAR");
+			break;
+		case SQLT_BLOB:
+			add_assoc_string(return_value, "oci:decl_type", "BLOB");
+			add_next_index_string(&flags, "blob");
+			add_assoc_string(return_value, "native_type", "BLOB");
+			break;
+		case SQLT_CLOB:
+			add_assoc_string(return_value, "oci:decl_type", "CLOB");
+			add_next_index_string(&flags, "blob");
+			add_assoc_string(return_value, "native_type", "CLOB");
+			break;
+		case SQLT_BFILE:
+			add_assoc_string(return_value, "oci:decl_type", "BFILE");
+			add_next_index_string(&flags, "blob");
+			add_assoc_string(return_value, "native_type", "BFILE");
+			break;
+		case SQLT_RDD:
+			add_assoc_string(return_value, "oci:decl_type", "ROWID");
+			add_assoc_string(return_value, "native_type", "ROWID");
+			break;
+		case SQLT_FLT :
+		case SQLT_BFLOAT:
+		case SQLT_IBFLOAT:
+			add_assoc_string(return_value, "oci:decl_type", "FLOAT");
+			add_assoc_string(return_value, "native_type", "FLOAT");
+			break;
+		case SQLT_BDOUBLE:
+		case SQLT_IBDOUBLE:
+			add_assoc_string(return_value, "oci:decl_type", "DOUBLE");
+			add_assoc_string(return_value, "native_type", "DOUBLE");
+			break;
+		default:
+			add_assoc_long(return_value, "oci:decl_type", dtype);
+			add_assoc_string(return_value, "native_type", "UNKNOWN");
+		}
+	} else if (data_size) {
+		/* if the column is the result of a function */
+		add_assoc_string(return_value, "native_type", "UNKNOWN");
+	} else {
+		/* if the column is NULL */
+		add_assoc_long(return_value, "oci:decl_type", 0);
+		add_assoc_string(return_value, "native_type", "NULL");
+	}
+
+	/* column can be null */
+	STMT_CALL_MSG(OCIAttrGet, "OCI_ATTR_IS_NULL",
+			(param, OCI_DTYPE_PARAM, &isnull, 0, OCI_ATTR_IS_NULL, S->err));
+
+	/* column name */
+	STMT_CALL_MSG(OCIAttrGet, "OCI_ATTR_NAME",
+			(param, OCI_DTYPE_PARAM, &colname, (ub4 *) &namelen, OCI_ATTR_NAME, S->err));
+
+	add_assoc_long(return_value, "precision", precis);
+	add_assoc_long(return_value, "len", data_size);
+	add_assoc_string(return_value, "name", (char *) colname);
+
+	if (isnull) {
+		add_next_index_string(&flags, "nullable");
+	} else {
+		add_next_index_string(&flags, "not_null");
+	}
+
+	/* PDO type */
+	switch (dtype) {
+		case SQLT_BFILE:
+		case SQLT_BLOB:
+		case SQLT_CLOB:
+			add_assoc_long(return_value, "pdo_type", PDO_PARAM_LOB);
+			break;
+		default:
+			add_assoc_long(return_value, "pdo_type", PDO_PARAM_STR);
+	}
+
+	add_assoc_zval(return_value, "flags", &flags);
+	OCIDescriptorFree(param, OCI_DTYPE_PARAM);
```

Comment:
Do we care that errors from calls like STMT_CALL_MSG will cause the function to return without freeing the descriptor?  It's a small leak in a rare (?) failure case.  Your call.

Reply:
My code is based on the oci_stmt_describe function which does the same things.
It doesn't free the OCIParam, so i don't know if it's even necessary as it may be free somewhere else.
Or there is also a memory leak problem in that function.